feature/text_based.py

```diff
@@ -0,0 +1,702 @@
+                cost = diversity_cost
+
+        # Compressed sparse column (transposed for convenience)
+        sparse_col = sparse.csr_matrix(self.matrix.T, copy=True)
+
+        # Initial guess of the Lagrangian multiplier with greedy algorithm
+        adjusted_cost = cost / sparse_col.dot(np.ones(self._num_rows))
+        cost_matrix = adjusted_cost * self.matrix + np.amax(adjusted_cost) * (~self.matrix)
+        u = adjusted_cost[np.argmin(cost_matrix, axis=1)]
+
+        # Nothing is selected, everything is uncovered
+        selected = np.zeros(self._num_cols, dtype=bool)
+        iuncovered = np.ones(self._num_rows, dtype=bool)
+        score = np.zeros(self._num_cols)
+
+        epsilon = 1E-5
+        size = sum(selected)
+
+        # While there are uncovered rows and below selection size
+        while np.count_nonzero(iuncovered) > 0 and size < selection_size:
+
+            # Faster than indexing, made possible by sparse_col.dot
+            mu = sparse_col.dot(iuncovered.astype(int)).astype(float)
+            mu[mu <= epsilon] = epsilon
+            # Set Lagrange multiplier zero for covered rows
+            u[~iuncovered] = 0
+            gamma = (cost - sparse_col.dot(u))
+            select_gamma = (gamma >= 0)
+
+            if np.count_nonzero(select_gamma) > 0:
+                score[select_gamma] = gamma[select_gamma] / mu[select_gamma]
+
+            if np.count_nonzero(~select_gamma) > 0:
+                score[~select_gamma] = gamma[~select_gamma] * mu[~select_gamma]
+
+            # Add new column (column with minimum cost that has not been selected yet
+            inewcolumn = (np.nonzero(~selected)[0])[np.argmin(score[~selected])]
+            selected[inewcolumn] = True
+            size += 1
+
+            iuncovered = ~np.logical_or(~iuncovered, self.matrix[:, inewcolumn])
+
+        if size == selection_size:
+            print("Warning: max greedy reached selection size", selection_size)
+
+        # Solution
+        selected = list(selected.nonzero()[0])
+        num_row_covered = self._get_num_row_covered(selected)
+
+        if self.verbose:
+            print("\nGREEDY SELECTION:", len(selected), "columns to cover rows ", self._num_rows)
+            print("=" * 40)
+            print("SIZE:", len(selected), "reduction: {:.2f}".format((self._num_cols - len(selected)) / self._num_cols))
+            print("SELECTED:", selected)
+            print("NUM ROWS COVERED:", num_row_covered, "coverage: {:.2f}".format(num_row_covered / self._num_rows))
+            print("STATUS: GREEDY")
+            print("COST METRIC:", self.cost_metric)
+            print("=" * 40)
+
+        return selected
+
+    def _select_random(self, trials: int = 10) -> List[int]:
+        # Set the seed
+        random.seed(self.seed)
+        best_selected = []
+        best_covered = 0
+
+        if self.selection_size is None:
+            self.selection_size = self._get_selection_size()
```

Comment:
_get_selection_size() returns a tuple so I'm not following logic in line 449. Did you intend to number of items selected (i.e., second value in tuple)?

Reply:
yeah this is an error (I made it)

Reply:
I fixed and you should be able to run all the tests.